File: priv/combined_metrics/samples/variable_naming/no_hungarian_notation/bad/inventory.rb
# Inventory management using Hungarian notation prefixes.
# BAD: str_name, b_active, n_count, arr_items, int_age, obj_user, fn_callback, d_price.

class InventoryBad
  def add_item(obj_item)
    str_name = obj_item[:name]
    str_sku = obj_item[:sku]
    n_quantity = obj_item[:quantity]
    d_price = obj_item[:price]
    b_active = obj_item.fetch(:active, true)

    return { ok: false, error: 'Name is required' } if str_name.nil? || str_name.strip.empty?

    {
      ok: true,
      data: { id: generate_id, name: str_name, sku: str_sku, quantity: n_quantity, price: d_price, active: b_active }
    }
  end

  def update_stock(str_item_id, n_delta)
    obj_item = fetch_item(str_item_id)
    n_new_quantity = obj_item[:quantity] + n_delta

    return { ok: false, error: 'Insufficient stock' } if n_new_quantity < 0

    { ok: true, data: obj_item.merge(quantity: n_new_quantity) }
  end

  def search_items(str_query, arr_items)
    str_lower_query = str_query.downcase

    arr_items.select do |obj_item|
      obj_item[:name].downcase.include?(str_lower_query) ||
        obj_item[:sku].downcase.include?(str_lower_query)
    end
  end

  def calculate_value(arr_items)
    arr_items.sum { |obj_item| obj_item[:quantity] * obj_item[:price] }
  end

  def apply_discount(arr_items, d_discount_rate, fn_filter)
    arr_items
      .select(&fn_filter)
      .map do |obj_item|
        d_new_price = (obj_item[:price] * (1 - d_discount_rate)).round(2)
        obj_item.merge(price: d_new_price)
      end
  end

  def group_by_category(arr_items)
    arr_items.group_by { |obj_item| obj_item[:category] }
  end

  def low_stock_report(arr_items, n_threshold)
    b_include_inactive = false

    arr_items
      .select { |obj_item| obj_item[:quantity] <= n_threshold && (b_include_inactive || obj_item[:active]) }
      .sort_by { |obj_item| obj_item[:quantity] }
  end

  def import_items(arr_raw_data, fn_transform, fn_validate)
    arr_transformed = arr_raw_data.map(&fn_transform)
    arr_valid = arr_transformed.select(&fn_validate)
    n_imported = arr_valid.length
    n_skipped = arr_raw_data.length - n_imported

    { imported: n_imported, skipped: n_skipped, items: arr_valid }
  end

  private

  def fetch_item(str_id)
    { id: str_id, name: 'Item', sku: 'SKU', quantity: 10, price: 9.99, active: true, category: 'misc' }
  end

  def generate_id = SecureRandom.hex(8)
end

File: priv/combined_metrics/samples/variable_naming/no_hungarian_notation/bad/inventory.js
// Inventory management using Hungarian notation prefixes.
// BAD: strName, bActive, nCount, arrItems, intAge, objUser, fnCallback, dPrice.

function addItem(objItem) {
  const strName = objItem.name;
  const strSku = objItem.sku;
  const nQuantity = objItem.quantity;
  const dPrice = objItem.price;
  const bActive = objItem.active ?? true;

  if (!strName || strName.trim() === '') {
    return { ok: false, error: 'Name is required' };
  }

  return {
    ok: true,
    data: { id: generateId(), name: strName, sku: strSku, quantity: nQuantity, price: dPrice, active: bActive },
  };
}

function updateStock(strItemId, nDelta) {
  const objItem = fetchItem(strItemId);
  const nNewQuantity = objItem.quantity + nDelta;

  if (nNewQuantity < 0) {
    return { ok: false, error: 'Insufficient stock' };
  }

  return { ok: true, data: { ...objItem, quantity: nNewQuantity } };
}

function searchItems(strQuery, arrItems) {
  const strLowerQuery = strQuery.toLowerCase();

  return arrItems.filter(objItem =>
    objItem.name.toLowerCase().includes(strLowerQuery) ||
    objItem.sku.toLowerCase().includes(strLowerQuery)
  );
}

function calculateValue(arrItems) {
  return arrItems.reduce((nAcc, objItem) => nAcc + objItem.quantity * objItem.price, 0);
}

function applyDiscount(arrItems, dDiscountRate, fnFilter) {
  return arrItems
    .filter(fnFilter)
    .map(objItem => {
      const dNewPrice = Math.round(objItem.price * (1 - dDiscountRate) * 100) / 100;
      return { ...objItem, price: dNewPrice };
    });
}

function groupByCategory(arrItems) {
  return arrItems.reduce((objAcc, objItem) => {
    const strCategory = objItem.category;
    objAcc[strCategory] = objAcc[strCategory] || [];
    objAcc[strCategory].push(objItem);
    return objAcc;
  }, {});
}

function lowStockReport(arrItems, nThreshold) {
  const bIncludeInactive = false;

  return arrItems
    .filter(objItem => objItem.quantity <= nThreshold && (bIncludeInactive || objItem.active))
    .sort((objA, objB) => objA.quantity - objB.quantity);
}

function importItems(arrRawData, fnTransform, fnValidate) {
  const arrTransformed = arrRawData.map(fnTransform);
  const arrValid = arrTransformed.filter(fnValidate);
  const nImported = arrValid.length;
  const nSkipped = arrRawData.length - nImported;

  return { imported: nImported, skipped: nSkipped, items: arrValid };
}

function fetchItem(strId) { return { id: strId, name: 'Item', sku: 'SKU', quantity: 10, price: 9.99, active: true, category: 'misc' }; }
function generateId() { return Math.random().toString(36).slice(2); }

module.exports = { addItem, updateStock, searchItems, calculateValue, applyDiscount, groupByCategory, lowStockReport, importItems };
